src/high-level/constructors.lisp

```diff
@@ -0,0 +1,177 @@
+;;;; constructors.lisp
+;;;;
+;;;; Author: Cole Scott
+
+(in-package #:magicl)
+
+(defvar *default-tensor-type* 'double-float
+  "The default element type to be used for constructing tensors when TYPE is not specified.")
+
+(defgeneric make-tensor (class shape &key initial-element layout storage)
+  (:documentation "Make a dense tensor with elements of the specified type"))
+
+(defun empty (shape &key (type *default-tensor-type*) layout)
+  "Create a tensor without intializing the contents of the storage
+
+If TYPE is not specified then *DEFAULT-TENSOR-TYPE* is used.
+LAYOUT specifies the internal storage representation ordering of the returned tensor.
+The tensor specialized on the specified SHAPE and TYPE."
+  (policy-cond:with-expectations (> speed safety)
+      ((type shape shape))
+    (let ((tensor-type (infer-tensor-type type shape nil)))
+      (make-tensor tensor-type shape :layout layout))))
+
+(defun const (const shape &key type layout)
+  "Create a tensor with the specified SHAPE with each element being set to CONST
+
+If TYPE is not specified then it is inferred from the type of CONST.
+LAYOUT specifies the internal storage representation ordering of the returned tensor.
+The tensor is specialized on SHAPE and TYPE."
+  (policy-cond:with-expectations (> speed safety)
+      ((type shape shape))
+    (let ((tensor-class (infer-tensor-type type shape const)))
+      (make-tensor tensor-class shape :layout layout :initial-element const))))
+
+(defun rand (shape &key (type *default-tensor-type*) layout distribution)
+  "Create tensor with random elements from DISTRIBUTION
+
+DISTRIBUTION is a function with no arguments which returns a value for the element.
+If DISTRIBUTION is not specified then a uniform distribution on [0,1] (or [0,1] + [0,1]i for complex types) is used.
+If TYPE is not specified then *DEFAULT-TENSOR-TYPE* is used.
+LAYOUT specifies the internal storage representation ordering of the returned tensor.
+The tensor is specialized on SHAPE and TYPE."
+  (policy-cond:with-expectations (> speed safety)
+      ((type shape shape))
+    (let* ((tensor-class (infer-tensor-type type shape nil))
+           (rand-function
+             (or distribution
+                 (cond
+                   ((subtypep type 'complex)
+                    (lambda ()
+                      (complex
+                       (random 1d0)
+                       (random 1d0))))
+                   (t
+                    (lambda ()
+                      (random 1d0))))))
+           (f (lambda (&rest rest)
+                (declare (ignore rest))
+                (coerce  (funcall rand-function) type))))
+      (into! f (make-tensor tensor-class shape :layout layout)))))
+
+(defun eye (shape &key value (offset 0) (type *default-tensor-type*) layout)
```

Comment:
I agree with having the integer argument yield a square matrix automatically. In general, taking a shape parameter doesn't seem outlandish to me. One could define a generalization of the Kronecker delta (but not the one from https://en.wikipedia.org/wiki/Kronecker_delta#Generalizations) such as:

    δ_{i_1, ..., i_r}^{j_1, ..., j_s} = 1 if i_1 = ... = i_r = j_1 = ... = j_s or 0 otherwise.

This is exactly what `numpy.eye` does. I can foresee it being useful when doing the SVD of a matrix and replacing the diagonal factor by another diagonal where all the singular values are one.

Reply:
I agree on this one. It might make sense to allow for either a shape to be specified or just a fixnum supplied for a square matrix.

Reply:
Would anyone have any strong opinions on letting `layout` be either a list of dimensions (what it is now) or a num which would then just build the corresponding square identity matrix?

Reply:
This seems to work on TENSORs, however for VECTORs it does _technically_ meet the requirement because the one dimension is equal to itself.

@jmbr Do you have any opinion on `(= (zeros '(2)) (eye '(2)))` vs `(= (ones '(2)) (eye '(2)))`

Reply:
I'm with @kilimanjaro. I'd like `(eye 2)` (note the absence of a list) to yield a 2x2 identity matrix by default.

Reply:
Let's go with taking `(eye '(2))` as a 2x2 identity matrix unless the user writes `(eye '(2 1))` in which case the column vector (1, 0) should be returned. This behavior would be consistent with both NumPy and MATLAB.

Reply:
Hmm. I think `(eye 2)` should (and does) return a 2x2 identity matrix but I think `(eye '(2))` should be a vector to be consistent with the rest of the API. I am not entirely sure if that returned vector should be just ones (it shouldn't), zeros, or only the first element being 1 with the rest being 0

Reply:
My opinion is that consistency with NumPy and MATLAB outweighs internal consistency within the API. However, if you want to go the route of taking that to be a vector, then it should have only the first element be 1 and the rest 0.

Reply:
If matlab/numpy consistency is desired, note that numpy always returns 2d arrays. So `np.eye(n)` returns an array of shape `(n,n)`, and `np.eye(n,m)` yields shape `(n, m)`. For us then, `eye` could take either an integer or a shape, so `(eye n)` is equivalent to `(eye (list n n))`. It should be an error to provide `eye` with a shape that is not a valid matrix shape, e.g. `(eye (list n))`.